strapi/strapi
# Upgrade to Apollo v4

### What does it do?

- Upgrade from Apollo v3 to Apollo v4

Some significant changes:

- [Apollo v4 has many breaking changes](https://www.apollographql.com/docs/apollo-server/migration/) such as server config options
- upgrade from `graphql` v15 to v16 [which has its own long list of breaking changes](https://github.com/graphql/graphql-js/releases/tag/v16.0.0)
- upgrade from `graphql-upload` v13 to v15 (note: didn't use v16 because it's esm only)
- Graphql Playground replaced with Apollo LandingLocal
  - Add `helmet` exceptions for it
  - Add service to check that landing page is enabled (so it can be detected outside the plugin, eg for core/security to configure helmet as needed)

- added some test utility functionality so request agent can send specific headers globally or per-request 

Follow-up in later PRs:

- [helmet exceptions on a per-route basis rather than for all special paths](https://github.com/strapi/strapi/pull/19202)
- investigate if the changes to Apollo v4 configuration and our current method of merging user config with default config have made anything 'impossible'; for example, `context` no longer exists in ApolloConfig and moved to the KoaMiddleware, so we may need to provide a mechanism for users to inject code and configuration there as well
- investigate what codemods can mitigate the breaking changes in user code
- rename all references to 'playground' since it's no longer accurate (`Playground` is EOL, only `LandingLocalDefault` is supported)
  - boolean 'playgroundAlways' will become strings `productionLandingPage` and `landingPage` with options 'LocalDefault' and 'ProductionDefault' (so that in the future we can then support other landing pages)
  - 'isPlaygroundEnabled' will then become 'landingPage' and return the current landing page being used
- add documentation to strapi/documentation for at least the following breaking changes:
  - all the package upgrades themselves
  - the need for the x-apollo-operation-name header in multipart requests
  - the replacement of playground

### Why is it needed?

Apollo v3 is already EOL, so Strapi v5 must support Apollo v4.

### How to test it?

- Going to /graphql should load the graphql landing page (which is now different than the old playground) and work (schemas loading, queries work, etc)
- The Graphql API should still work the same as before
- The graphql server within Strapi should still be accessible as before and user code should generally work, as long as it's not using Apollo v3 specific functionality
- API tests should all still pass
- try changing some plugins.js apollo configuration settings -- all the ones that weren't breaking changes should still work

### Related issue(s)/PR(s)

#19202 - First follow-up on this to improve the helmet security config
